24.md

```diff
@@ -17,7 +17,32 @@ These are extra fields not specified in NIP-01 that may be present in the string
   - `website`: a web URL related in any way to the event author.
   - `banner`: an URL to a wide (~1024x768) picture to be optionally displayed in the background of a profile screen.
   - `bot`: a boolean to clarify that the content is entirely or partially the result of automation, such as with chatbots or newsfeeds.
-  - `pronouns`: a string representing the preferred way to refer to this person.
+
+### User defined custom fields
+
+A `fields` attribute may also be added to profile metadata to represent user-defined fields. It is an array name-value tuples, eg:
+
+```json
+[["<name>", "<value>"], ...]
+```
+
+In the context of a full metadata object:
+
+```json
+{
+  "name": "Alex",
+  "avatar": "https://...",
+  "fields": [
+    ["Pronouns", "ye/haw"],
+    ["Lifestyle", "vegan"],
+    ["Color", "green"]
+  ]
```

Comment:
why have extra fields, and not just put them in the JSON, as JSON is extensible by design?

Reply:
Because of this:

> While protocol-level fields like `lud16` and `picture` can influence the behavior of a user's account, custom fields are used for display purposes only. Clients should display custom fields in a confined area of the user's profile, and not make assumptions about their meaning.

Reply:
ACK.  This seems a pragmatic compromise.

Reply:
I've already allowed creating/displaying custom fields by just adding more key-value pairs in the content.... back in 2022.

Is there really a problem with metadata fields being interpreted and used by software that thinks it knows what they mean, colliding with people using that same field just for display purposes?

Reply:
> Is there really a problem with metadata fields [...] colliding

I think so. I wouldn't want the user to directly modify fields like lud16 and nip05 without validation in my client. It sounds like what you have is a protocol-level advanced feature for power users, not Custom Profile Fields. Custom fields can have duplicate entries and are displayed in a user-defined order.

Reply:
The only side effect of @mikedilger's model is that some properties on kind 0 are not to be displayed. They could be just for internal use. The array clarifies the order these fields must be shown and that they are indeed supposed to be shown.

Reply:
If you throw them in the same bag as the standardized fields, then if the user defines a custom field that has an existing meaning, or that gains a standardized meaning later, that can give unexpected behavior.. Clients have to check for this, which adds extra complexity as well.

With this syntax it's obvious what fields to show in the table and in what order.

Reply:
> or that gains a standardized meaning later, that can give unexpected behavior

For example what if, in your client, the user has added a custom field named "fields". After this NIP, those kind 0 events will become invalid, because the type should be an array not a string. That's why it's bad to mix something fully user controlled with protocol fields.